### PR TITLE
testthat 3.2.0

### PR DESCRIPTION
**Pull request by hadley**
Are there other cool new features that I've forgotten?

### CODE REVIEW

File: content/blog/testthat-3-2-0/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+* [ ] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
+* [x] Find photo & update yaml metadata
+* [x] Create `thumbnail-sq.jpg`; height and width should be equal
+* [x] Create `thumbnail-wd.jpg`; width should be >5x height
+* [x] `hugodown::use_tidy_thumbnails()`
+* [x] Add intro sentence, e.g. the standard tagline for the package
+* [x] `usethis::use_tidy_thanks()`
+-->
+```
+We're chuffed to announce the release of [testthat](%7B%20home%20%7D) 3.2.0.
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:37]: Link issue?

File: content/blog/testthat-3-2-0/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+testthat makes it easy to turn your existing informal tests into formal, automated tests that you can rerun quickly and easily.
+testthat is the most popular unit-testing package for R, and is used by almost 9,000 CRAN and Bioconductor packages.
+You can learn more about unit testing at [https://r-pkgs.org/tests.html](#0).
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:40]: Link issue?

File: content/blog/testthat-3-2-0/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+Here we'll focus on the biggest news: new expectations, tweaks to the way that error snapshots are reported, a new way to detect if a test has changed global state, and a handful of UI improvements.
+
+```{r setup}
+library(testthat)
+```
+
+## New expectations
+
+New `expect_contains()` and `expect_in()` that works similarly to `expect_true(all(expected %in% object))` or `expect_true(all(object %in% expected))` but give more informative failure messages:
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:60]:
```suggestion
New `expect_contains()` and `expect_in()` that work similarly to `expect_true(all(expected %in% object))` or `expect_true(all(object %in% expected))` but give more informative failure messages:
```

File: content/blog/testthat-3-2-0/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+expect_contains(fruits, "apple")
+expect_contains(fruits, "pineapple")
+
+x <- c(TRUE, FALSE, TRUE, FALSE)
+expect_in(x, c(TRUE, FALSE))
+x <- c(TRUE, FALSE, TRUE, NA, FALSE)
+expect_in(x, c(TRUE, FALSE))
+```
+
+New `expect_no_error()`, `expect_no_warning()`, and `expect_no_message()` make it easier (and clearer) to expect that code runs with without errors, warnings, or messages.
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:73]:
```suggestion
New `expect_no_error()`, `expect_no_warning()`, and `expect_no_message()` make it easier (and clearer) to expect that code runs without errors, warnings, or messages.
```

File: content/blog/testthat-3-2-0/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  )
+})
+```
+
+When you look at a single test it's easy to spot the problem, and switch to a more appropriate way of temporarily changing the options, like `withr::local_options()`.
+But sometimes this mistake crept in a long time ago and is now hiding amongst hundreds or thousands of tests.
+
+In earlier versions of testthat, finding tests that accidentally changed the world was painful: the only way was to painstakingly review each test.
+Now you can use `set_state_inspector()` to register a function that's called before and after every test.
+The function returns different values, testthat will let you know.
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:107]:
```suggestion
If the function returns different values, testthat will let you know.
```

File: content/blog/testthat-3-2-0/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+So, for example, to detect if any of your tests have modified options you could put the following in `tests/testhat/setup.R` or an existing helper file:
+
+```{r}
+set_state_inspector(function() {
+  list(options = options())
+})
+```
+
+Or maybe you've an `R CMD check` warning that you've forgotten to close a connection:
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:117]:
```suggestion
Or maybe you've seen an `R CMD check` warning that you've forgotten to close a connection:
```

File: content/blog/testthat-3-2-0/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+set_state_inspector(function() {
+  list(connections = nrow(showConnections()))
+})
+```
+
+And you can of course combine multiple checks just by returning a more complicated list.
+
+## UI improvements
+
+testthat 3.2.0 includes a bunch of minor user interface improvements that should make day-to-day use of testthat more enjoyable.
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:129]: I faintly recall you saying that you believe parallel testing is now much more robust, do you want to mention it?

File: content/blog/testthat-3-2-0/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+Some of our favourite highlights are:
+
+-   We use an improved algorithm to find the source reference associated with an expectation/error/warning/skip.
+    We now look for the most recent call (within inside `test_that()` that has known source.
+    This generally gives more specific locations than the previous approach and gives much better locations if an error occurs in an exit handler.
+
+-   Tracebacks are no longer truncated we uses rlang's default tree display; this should make it easier to track down problems when testing in non-interactive contexts.
```
Comment (DavisVaughan) [inline content/blog/testthat-3-2-0/index.Rmd:136]:
```suggestion
-   Tracebacks are no longer truncated and we use rlang's default tree display; this should make it easier to track down problems when testing in non-interactive contexts.
```

**Review by DavisVaughan — APPROVED** [review]

File: content/blog/testthat-3-2-0/index.Rmd [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    waldo::compare(c("X", letters), c(letters, "X"))
+  )
+})
+```
+
+When you look at a single test it's easy to spot the problem, and switch to a more appropriate way of temporarily changing the options, like `withr::local_options()`.
+But sometimes this mistake crept in a long time ago and is now hiding amongst hundreds or thousands of tests.
+
+In earlier versions of testthat, finding tests that accidentally changed the world was painful: the only way was to painstakingly review each test.
+Now you can use `set_state_inspector()` to register a function that's called before and after every test.
```
Comment (jennybc) [inline content/blog/testthat-3-2-0/index.Rmd:106]: Do you want to say where to put the call to `set_state_inspector()`, i.e. in `tests/testthat/setup.R`?

**Review by jennybc — APPROVED** [review]
Do you want to mention the ability to run a single test? No strong feelings here, just wanted to raise it.

Maybe mention the new vignette that explains important files? `vignette("special-files")`

Mention expanded coverage of testing in R Packages 2e?

Mention the return of mocking?

**Comment by hadley** [comment]
I can't believe I forgot about mocking; that's the whole reason I was thinking of this post!

File: content/blog/testthat-3-2-0/index.Rmd (hunk 7 above)
Comment (gaborcsardi) [reply content/blog/testthat-3-2-0/index.Rmd:129]: Indeed there weren't any changes in parallel testing from the first implementation AFAIR, except that it interacts better with snapshots now, and testthat deletes unneeded snapshots files.

Comment (kevinushey) [reply content/blog/testthat-3-2-0/index.Rmd:129]: I think the main parallel-related fix happened outside of testthat? IIRC there was an issue with processes sometimes getting stuck / left behind if you terminated `devtools::test()` while it was still trying to run tests.

Comment (hadley) [reply content/blog/testthat-3-2-0/index.Rmd:220]: Mentioned both points.

File: content/blog/testthat-3-2-0/index.Rmd [hunk 10]
```diff
@@ -70,21 +84,97 @@ x <- c(TRUE, FALSE, TRUE, NA, FALSE)
...
+  b <- g(y, z)
+  c <- h(a, b)
+  
+  d <- complicated(c)
+  
+  i(d, 1, TRUE)
+}
+```
+
+Mocking allows you to temporarily `complicated()` with something simpler, allowing you to test the rest of the function.
```
Comment (juliasilge) [inline content/blog/testthat-3-2-0/index.Rmd:157]:
```suggestion
Mocking allows you to temporarily replace `complicated()` with something simpler, allowing you to test the rest of the function.
```

**Comment by hadley** [comment]
* [ ] Update date
* [ ] Rebuild file

File: content/blog/testthat-3-2-0/index.Rmd [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+## Documentation
+
+The first and most important thing to point out is that the second edition of [R Packages](https://r-pkgs.org) contains updated and much expanded coverage of testing.
+Coverage of testing is now split up over three chapters:
+
+-   [Testing basics](https://r-pkgs.org/testing-basics.html)
+-   [Designing your test suite](https://r-pkgs.org/testing-design.html)
+-   [Advanced testing techniques](https://r-pkgs.org/testing-advanced.html)
+
+There's also a new vignette about special files (`vignette("special-files")`) which describes the various special files that you find in `tests/testthat` and when you might need to use them.
```
Comment (jennybc) [inline content/blog/testthat-3-2-0/index.Rmd:68]: The vignette isn't auto-linking. Does it need the package to be specified?

File: content/blog/testthat-3-2-0/index.Rmd [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+expect_contains(fruits, "pineapple")
+
+x <- c(TRUE, FALSE, TRUE, FALSE)
+expect_in(x, c(TRUE, FALSE))
+x <- c(TRUE, FALSE, TRUE, NA, FALSE)
+expect_in(x, c(TRUE, FALSE))
+```
+
+`expect_no_error()`, `expect_no_warning()`, and `expect_no_message()` make it easier (and clearer) to confirm that code runs without errors, warnings, or messages.
+The default fails if there is any error/warning/message, but you can optionally supply either the `message` or `class` arguments to confirm that a specific error/warning/message is no longer present.
```
Comment (jennybc) [inline content/blog/testthat-3-2-0/index.Rmd:88]:
```suggestion
The default fails if there is any error/warning/message, but you can optionally supply either the `message` or `class` arguments to confirm the absence of a specific error/warning/message.
```

**Review by jennybc — APPROVED** [review]

File: content/blog/testthat-3-2-0/index.Rmd (hunk 11 above)
Comment (hadley) [reply content/blog/testthat-3-2-0/index.Rmd:68]: Probably not linking because isn't on CRAN yet? Hopefully will momentarily.

Comment (hadley) [reply content/blog/testthat-3-2-0/index.Rmd:68]: Yes!

**Pull request merged**